File: bindings/dart/bao/lib/src/fileinfo.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:bao/bao.dart';

class FileInfo {
  late int id;
  late String name;
  late int size;
  late int allocatedSize;
  late DateTime modTime;
  late bool isDir;
  late int flags;
  late Uint8List attrs;
  late int keyId;
  late String storageDir;
  late String storageName;
  late PublicID authorId;
  static var none = FileInfo('', 0, DateTime.fromMillisecondsSinceEpoch(0), false, 0);

  FileInfo(this.name, this.size, this.modTime, this.isDir, this.id);

  FileInfo.fromMap(Map<String, dynamic> map) {
    id = map['id'] ?? 0;
    name = map['name'];
    size = map['size'];
    modTime = DateTime.parse(map['modTime']);
    isDir = map['isDir'];
    flags = map['flags'] ?? 0;
    allocatedSize = map['allocatedSize'] ?? 0;
    // attrs is a Go []byte; when JSON-encoded it becomes a base64 string. Decode to bytes.
    final a = map['attrs'];
    if (a == null) {
      attrs = Uint8List(0);
    } else if (a is String) {
      attrs = Uint8List.fromList(base64Decode(a));
    } else if (a is List<int>) {
      attrs = Uint8List.fromList(a);
    } else if (a is Uint8List) {
      attrs = a;
    } else {
      attrs = Uint8List(0);
    }
    keyId = map['keyId'] ?? 0;
    storageDir = map['storageDir'] ?? '';
    storageName = map['storageName'] ?? '';
    authorId = map['authorId'] ?? '';
  }
}

File: bindings/java/lib/src/main/java/ink/francesco/bao/FileInfo.java
package ink.francesco.bao;

import java.time.Instant;
import java.util.Base64;
import java.util.Map;

public class FileInfo {
    public long id;
    public String name;
    public long size;
    public long allocatedSize;
    public Instant modTime;
    public boolean isDir;
    public long flags;
    public byte[] attrs;
    public long keyId;
    public String storageDir;
    public String storageName;
    public String authorId;

    static FileInfo fromMap(Map<String, Object> m) {
        var fi = new FileInfo();
        fi.id = ((Number) m.getOrDefault("id", 0)).longValue();
        fi.name = (String) m.getOrDefault("name", "");
        fi.size = ((Number) m.getOrDefault("size", 0)).longValue();
        fi.allocatedSize = ((Number) m.getOrDefault("allocatedSize", 0)).longValue();
        var mod = (String) m.getOrDefault("modTime", "1970-01-01T00:00:00Z");
        fi.modTime = Instant.parse(mod);
        fi.isDir = Boolean.TRUE.equals(m.get("isDir"));
        fi.flags = ((Number) m.getOrDefault("flags", 0)).longValue();

        Object a = m.get("attrs");
        if (a instanceof String s) {
            fi.attrs = Base64.getDecoder().decode(s);
        } else if (a instanceof byte[] b) {
            fi.attrs = b;
        } else {
            fi.attrs = new byte[0];
        }
        fi.keyId = ((Number) m.getOrDefault("keyId", 0)).longValue();
        fi.storageDir = (String) m.getOrDefault("storageDir", "");
        fi.storageName = (String) m.getOrDefault("storageName", "");
        fi.authorId = (String) m.getOrDefault("authorId", "");
        return fi;
    }
}
